src/snowflake/snowpark/session.py

```diff
@@ -1149,27 +1189,161 @@ def _upload_unsupported_packages(
+                metadata[environment_signature] = "|".join(
+                    [
+                        str(requirement)
+                        for requirement in supported_dependencies + new_dependencies
+                    ]
+                )
+                metadata_local_path = os.path.join(
+                    self._tmpdir_handler.name, metadata_file
+                )
+                with open(metadata_local_path, "w") as file:
+                    for key, value in metadata.items():
+                        file.write(f"{key},{value}\n")
+
+                # Upload metadata file to stage
+                # Note that the metadata file is not compressed, only the zip files are.
+                self._conn.upload_file(
+                    path=normalize_local_file(metadata_local_path),
+                    stage_location=normalize_remote_file_or_dir(stage_name),
+                    compress_data=False,
+                    overwrite=True,
+                )
+
+            self._conn.upload_file(
+                path=normalize_local_file(zip_path),
+                stage_location=normalize_remote_file_or_dir(stage_name),
+                compress_data=True,
+                overwrite=True,
+            )
+
+            # Add zipped file as an import
+            stage_zip_path = f"{stage_name}/{zip_file}"
+            self.add_import(
+                stage_zip_path
+                if stage_zip_path.startswith(STAGE_PREFIX)
+                else f"{STAGE_PREFIX}{stage_zip_path}"
+            )
         except Exception as e:
-            if self._tmpdir_handler:
-                self._tmpdir_handler.cleanup()
             raise RuntimeError(
                 f"Unable to auto-upload packages: {packages}, Error: {e} | NOTE: Alternatively, you can find the "
                 f"directory of these packages and add it via Session.add_import. See details at "
                 f"https://docs.snowflake.com/en/developer-guide/snowpark/python/creating-udfs.html#using"
                 f"-third-party-packages-from-anaconda-in-a-udf."
             )
+        finally:
+            if self._tmpdir_handler:
+                self._tmpdir_handler.cleanup()
+                self._tmpdir_handler = None
 
         return supported_dependencies + new_dependencies
 
     def _is_anaconda_terms_acknowledged(self) -> bool:
         return self._run_query("select system$are_anaconda_terms_acknowledged()")[0][0]
 
+    def _load_unsupported_packages_from_stage(
+        self, environment_signature: str
+    ) -> Optional[List[pkg_resources.Requirement]]:
+        """
+        Uses specified stage path to auto-import a group of unsupported packages, along with its dependencies. This
+        saves time spent on pip install, native package detection and zip upload to stage.
+
+        A cached environment on a stage consists of 2 files:
```

Comment:
I'd use "two files" here. Use a numeral when the quantity is more than nine or when the number is used with a measurement, as in "2 centimeters".